corehq/apps/app_manager/views/utils.py

```diff
@@ -167,12 +167,31 @@ def overwrite_app(app, master_build, report_map=None):
     ])
     master_json = master_build.to_json()
     app_json = app.to_json()
-    form_ids_by_xmlns = _get_form_ids_by_xmlns(app_json)  # do this before we change the source
+
+    # Corresponding forms in a master app and linked app need to have the same XMLNS but different unique ids,
+    # so the linked app needs to know if there are any new forms and, if so, assign those forms new unique ids.
+    # To do this lookup, get the XMLNSes from the the most recent versions of this app pulled from each master
+    # and compare those to the XMLNSes present in this app.
+    unknown_xmlnses = {form.xmlns for form in app.get_forms()}
+    form_ids_by_xmlns = {}
+    unknown_xmlnses = unknown_xmlnses.difference(form_ids_by_xmlns.keys())
+    for brief in app.get_master_app_briefs():
+        if len(unknown_xmlnses):
+            previous_app = app.get_latest_build_from_upstream(brief.master_id)
+            if previous_app:
+                form_ids_by_xmlns.update(_get_form_ids_by_xmlns(previous_app))
+                unknown_xmlnses = unknown_xmlnses.difference(form_ids_by_xmlns.keys())
+    # Add in any forms from the current linked app, before the source is overwritten.
+    # This is particularly important if there's no previous version.
+    if len(unknown_xmlnses):
+        for module in app['modules']:
+            for form in module['forms']:
+                form_ids_by_xmlns[form.xmlns] = form['unique_id']
```

Comment:
nit: this loop could be `form_ids_by_xmlns.update(_get_form_ids_by_xmlns(app))`

Reply:
e5ac4ba227d